js-apps/meep-frontend/src/js/containers/exec/scenario-update-event-pane.js

```diff
@@ -0,0 +1,428 @@
+        SCENARIO_UPDATE_ACTION_NONE,
+        SCENARIO_UPDATE_ACTION_ADD,
+        // SCENARIO_UPDATE_ACTION_MODIFY,
+        SCENARIO_UPDATE_ACTION_REMOVE
+      ]
+    };
+  }
+
+  componentDidMount() {
+    this.props.changeActionType(SCENARIO_UPDATE_ACTION_NONE);
+    this.props.execElemNew();
+    this.props.changeRemoveActionEleType('');
+    this.props.changeRemoveActionEleName('');
+  }
+
+  changeAction(action) {
+    this.props.changeActionType(action);
+    if (action === SCENARIO_UPDATE_ACTION_ADD) {
+      this.props.execElemNew();
+    } else if (action === SCENARIO_UPDATE_ACTION_REMOVE) {
+      this.props.changeRemoveActionEleType('');
+      this.props.changeRemoveActionEleName('');
+    }
+  }
+
+  changeElementType(elementType) {
+    this.props.changeRemoveActionEleType(elementType);
+    this.getElementNames(elementType, this.props.scenario);
+    this.props.changeRemoveActionEleName('');
+  }
+
+  getElementNames(elementName, scenario) {
+    elementNames.length = 0;
+    var neType = '';
+    switch(elementName) {
+    case ELEMENT_TYPE_UE:
+      neType = UE_TYPE_STR;
+      break;
+    case ELEMENT_TYPE_FOG:
+      neType = FOG_TYPE_STR;
+      break;
+    case ELEMENT_TYPE_EDGE:
+      neType = EDGE_APP_TYPE_STR;
+      break;
+    case ELEMENT_TYPE_DC:
+      neType = DC_TYPE_STR;
+      break;
+    case ELEMENT_TYPE_UE_APP:
+      neType = UE_APP_TYPE_STR;
+      break;
+    case ELEMENT_TYPE_EDGE_APP:
+      neType = EDGE_APP_TYPE_STR;
+      break;
+    case ELEMENT_TYPE_CLOUD_APP:
+      neType = CLOUD_APP_TYPE_STR;
+      break;
+    default:
+      return;
+    }
+    for (var dInd in scenario.deployment.domains) {
+      var domain = scenario.deployment.domains[dInd];
+      for (var zInd in domain.zones) {
+        var zone = domain.zones[zInd];
+        for (var nInd in zone.networkLocations) {
+          var nl = zone.networkLocations[nInd];
+          for (var plInd in nl.physicalLocations) {
+            var pl = nl.physicalLocations[plInd];
+            if (pl.type === neType) {
+              elementNames.push(pl.name);
+            }
+            for (var prInd in pl.processes) {
+              var pr = pl.processes[prInd];
+              if (pr.type === neType) {
+                elementNames.push(pr.name);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  onSaveElement(element) {
+    if (!validateNetworkElement(element, this.props.table.entries, this.props.execElemSetErrMsg)) {
+      return;
+    }
+
+    var action = '';
+    var updatedScenario = updateObject({}, this.props.scenario);
+    if (this.props.execConfigMode === EXEC_ELEM_MODE_NEW) {
+      addElementToScenario(updatedScenario, element);
+      action = SCENARIO_UPDATE_ACTION_ADD;
+    } else {
+      updateElementInScenario(updatedScenario, element);
+      action = SCENARIO_UPDATE_ACTION_MODIFY;
+    }
+
+    var pl = this.getPLFromScenario(getElemFieldVal(element, FIELD_NAME), updatedScenario);
```

Comment:
We should not edit the Exec page scenario here. We simply need to send a SCENARIO-UPDATE event with the new element. Once the backend receives and processes the update then it will take care of updating the active scenario. The Exec page polls this active scenario every 1 second and will get & process the updated scenario on its next poll.

Also, we should not parse the scenario here. The parsing is done every second and populates the element table. Is there some information missing that absolutely requires parsing the scenario again here?

Reply:
Updated this code